mice-and-cats-java/src/main/java/foop/world/World.java

```diff
@@ -65,46 +73,64 @@ public void afterEntityMoved(HashSet<Player> players) {
             return;
         }
 
-        var cat = entities.get(0);
+
+        var cats = entities.stream().filter(entity -> entity.getType() == CAT).toList();
         for (Entity e : entities) {
-            if (cat != e && !e.isDead() && cat.getPosition().equals(e.getPosition()) && cat.isUnderground() == e.isUnderground()) {
+            if (!cats.contains(e)
+                    && cats.stream().map(Entity::isUnderground).anyMatch(c -> c == e.isUnderground())
+                    && !e.isDead()
+                    && cats.stream().map(Entity::getPosition).anyMatch(c -> c.equals(e.getPosition()))) {
+                // entity was killed by the cat...
                 e.setDead(true);
-                var update = new EntityUpdateMessage(e.getId(), e.getName(), e.getPosition(), e.isUnderground(), e.isDead());
+                broadcastMsg(players, new EntityUpdateMessage(e));
                 // we copy to avoid iterator invalidation, when a player is removed
-                for (var player : new ArrayList<>(players)) {
-                    player.send(update);
-                    if (player.getName().equals(e.getName())) {
-                        player.gameOver(new GameOverMessage(GameOverMessage.Result.YOU_DIED));
-                    }
+
+
+                var tmp = players.stream().filter(player -> player.getName().equals(e.getName())).toList();
+                for (Player p : tmp) {
+                    p.gameOver(new GameOverMessage(GameOverMessage.Result.YOU_DIED));
                 }
```

Comment:
We (stupidly) assume the player name is unique, so findFirst would be fine as well